README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

Simply extract required values from specific paths in a `Map` or a `List`, for now it works for strings keys


## Usage

```dart
import 'dart:convert';

import 'package:path_selector/path_selector.dart';

void main() {
  final Map<String, dynamic> map = jsonDecode('''
    {
      "name": "Mohamed Al Ashaal",
      "age": 28,
      "contacts": {
        "email": "m7medalash3al@gmail.com",
        "facebook": "https://fb.me/alash3al",
        "github": "https://github.com/alash3al"
      },
      "some.dot-key.included": "Works!",
      "skills": [
        {
          "name": "PHP",
          "rating": 0.9
        },
        {
          "name": "Golang",
          "rating": 0.9
        },
        {
          "name": "Python",
          "rating": 0.9
        }
      ]
    }
    ''');

  // will print: Mohamed Al Ashaal
  print(map.select("name"));

  // will print: 20
  print(map.select("age"));

  // will print: m7medalash3al@gmail.com
  print(map.select("contacts.email"));

  // will print: [{name: PHP, rating: 0.9}, {name: Golang, rating: 0.9}, {name: Python, rating: 0.9}]
  print(map.select("skills"));

  // will print {name: PHP, rating: 0.9}
  print(map.select("skills.#0"));

  // will print: PHP
  print(map.select("skills.#0.name"));

  // will print: null
  print(map.select("skills2.unknown_key.value"));

  // will print: Works!
  print(map.select("some\\.dot-key\\.included"));
}
```


CHANGELOG.md
## 1.0.0

- Initial version.

## 1.3.0

- Allow dot in keys by using `\.` for escaping.

test/path_selector_test.dart
import 'dart:convert';

import 'package:path_selector/path_selector.dart';
import 'package:test/test.dart';

void main() {
  group('Basic path_selector test cases', () {
    final Map<String, dynamic> map = jsonDecode('''
    {
      "name": "Some Name",
      "age": 28,
      "some.dot-key.included": "Works!",
      "skills": [
        {
          "name": "PHP",
          "rating": 0.9
        },
        {
          "name": "Golang",
          "rating": 0.9
        },
        {
          "name": "Python",
          "rating": 0.9
        }
      ]
    }
    ''');

    final List<dynamic> list = jsonDecode('["A", "B", {"key": "value"}]');

    test('Can select from a Map?', () {
      expect(map.select<String>("name"), "Some Name");
    });
    test('Can nest-select from a Map?', () {
      expect(map.select<String>("skills.#0.name"), "PHP");
    });
    test('Can select from a List?', () {
      expect(list.select<String>("#0"), "A");
    });
    test('Can nest-select from a List?', () {
      expect(list.select<String>("#2.key"), "value");
    });
    test('Can select dotted key?', () {
      expect(map.select<String>("some\\.dot-key\\.included"), "Works!");
    });
  });
}


lib/src/path_selector_base.dart
/// extract the value at the specified path "dot-seprated string" from the specified data
T? selectFrom<T>(dynamic data, String path) {
  var parts = [];
  var currentKeyName = "";

  for (var i = 0; i < path.length; i++) {
    if (path[i] != "\\" && path[i] != ".") {
      currentKeyName += path[i];
      continue;
    }

    var hasAcharSkip = false;
    if (path[i] == "\\") {
      hasAcharSkip = true;
      i++;
    }

    if (path[i] == ".") {
      if (hasAcharSkip) {
        currentKeyName += path[i];
        continue;
      }

      parts.add(currentKeyName);
      currentKeyName = "";
      continue;
    }
  }

  if (currentKeyName != "") {
    parts.add(currentKeyName);
  }

  dynamic val = data;

  for (var i = 0; i < parts.length; i++) {
    if (parts[i].startsWith("#")) {
      int index = int.parse(parts[i].substring(1));
      val = (val as List<dynamic>)[index];
    } else {
      val = (val as Map<String, dynamic>)[parts[i]];
    }

    if (val == null) {
      return null;
    }
  }

  return val as T;
}

/// add a select method to any map
extension MapPathSelector<K, V> on Map<K, V> {
  T? select<T>(String path) {
    return selectFrom(this, path);
  }
}

/// add a select method to any list
extension ListPathSelector<V> on List<V> {
  T? select<T>(String path) {
    return selectFrom(this, path);
  }
}


example/path_selector_example.dart
import 'dart:convert';

import 'package:path_selector/path_selector.dart';

void main() {
  final Map<String, dynamic> map = jsonDecode('''
    {
      "name": "Mohamed Al Ashaal",
      "age": 28,
      "contacts": {
        "email": "m7medalash3al@gmail.com",
        "facebook": "https://fb.me/alash3al",
        "github": "https://github.com/alash3al"
      },
      "some.dot-key.included": "Works!",
      "skills": [
        {
          "name": "PHP",
          "rating": 0.9
        },
        {
          "name": "Golang",
          "rating": 0.9
        },
        {
          "name": "Python",
          "rating": 0.9
        }
      ]
    }
    ''');

  // will print: Mohamed Al Ashaal
  print(map.select("name"));

  // will print: 20
  print(map.select("age"));

  // will print: m7medalash3al@gmail.com
  print(map.select("contacts.email"));

  // will print: [{name: PHP, rating: 0.9}, {name: Golang, rating: 0.9}, {name: Python, rating: 0.9}]
  print(map.select("skills"));

  // will print {name: PHP, rating: 0.9}
  print(map.select("skills.#0"));

  // will print: PHP
  print(map.select("skills.#0.name"));

  // will print: null
  print(map.select("skills2.unknown_key.value"));

  // will print: Works!
  print(map.select("some\\.dot-key\\.included"));
}
